### PR TITLE
Javascript self vs bind(this)

### PR DESCRIPTION
**Pull request by lizcross**
# Self vs bind(this) which do we prefer

Status: **Ready for Review**
Reviewers: @jansepar all js devs

## Changes:

- Took the stance that we prefer self over bind(this)
- Discuss!

### CODE REVIEW

**Comment by johnboxall** [comment]
By analyzing the blame information on this pull request, we identified @scalvert, @ry5n and @tedtate to be potential reviewers.

**Comment by jvoll** [comment]
I am personally in favour of `var self = this` -- I find that I miss the .bind at the end of the block and that it is something I don't notice until I'm done reading the whole block at which point I have to reread/rethink the meaning of the block now knowing that the `bind` is present.

File: javascript/README.md [hunk 1]
```diff
@@ -402,3 +402,48 @@ Prefer the short, more succinct version.
 Array.prototype.forEach.call(arguments, function(arg) { console.log(arg); });
 ```
 
+##Prefer `self` to `bind(this)`
+
```
Comment (jeremywiebe) [inline javascript/README.md]: pew pew

File: javascript/README.md [hunk 2]
```diff
@@ -402,3 +402,48 @@ Prefer the short, more succinct version.
 Array.prototype.forEach.call(arguments, function(arg) { console.log(arg); });
 ```
 
+##Prefer `self` to `bind(this)`
```
Comment (jeremywiebe) [inline javascript/README.md]: Should we be more strong with the wording here?  "Always use" vs "Prefer".  Can anyone think of a reason to use `.bind()` where `self` wouldn't be just as easy/good?

**Comment by jansepar** [comment]
This stack overflow answer has some interesting points:

http://codereview.stackexchange.com/questions/49872/using-var-self-this-or-bindthis

One that I thought was interesting - using `bind` takes away your ability to grab the original `this` that would have been in scope without the bind.

**Comment by jansepar** [comment]
In general I think my vote would be :+1: to `var self = this`, unless anyone can sway me the other way :)

**Comment by jansepar** [comment]
Also, I'd say neither approach is necessarily *wrong*, I would just like to standardize / bring some consistency to how we do it, because right now how we do it is dependant on the individual.

File: javascript/README.md (hunk 2 above)
Comment (lizcross) [inline javascript/README.md]: I was just looking at some code I wrote. I wanted to pass in a prototype function into another function to be called at a later time. I think this is reasonable time to use `bind(this)`. Here is the line: https://github.com/mobify/astro-scaffold/blob/hyb-578-update-scaffold-header-bar-tabs/app/scaffold-controllers/navigationController.js#L38

**Comment by DanielBrookRoberge** [comment]
I'm :+1: on this, there's some conceptual tidiness with `bind` but it's way too unobtrusive.

**Comment by tedtate** [comment]
This is a bit of a side benefit of `self` over `bind` that I ran into recently when writing some tests:

```
var mySpy = sinon.spy(object, 'doThing');
$(window).on('event', object.doThing.bind(this));
```

If you write a test that uses spy/mock you won't be able to test the amount of times `doThing` is called due to it being a new method.

**Comment by jansepar** [comment]
Thx for the feedback @tedtate :)

At this point I'm happy to :+1: - doesn't seem like there are any objections for using .bind by default.

@lizcross :ship: :shipit:

**Comment by ericmuyser** [comment]
@tedtate Kind of curious, what if you did `object.doThing = object.doThing.bind(this)` before spy?

I know that would permanently change `object.doThing`, so you may want to unbind. I figure internally the interpreter is it's doing something similar to this anyway (capturing the scope to get 'self' and creating a closure). Thoughts?

**Comment by tedtate** [comment]
@ericmuyser I probably should have written that out a little clearer:

```
class Obj {
    constructor():
        $(window).on('event', object.doThing.bind(this));
}
```

```
test():
    var object = new Obj();
    var mySpy = sinon.spy(object, 'doThing');
```

Spy creation happens after the fact so you don't really have that option.

**Comment by ericmuyser** [comment]
Ah, that totally makes sense. Thanks @tedtate !

I've never seen anybody do this before, but now I'm wondering what the downside to [doing this would be](https://jsfiddle.net/jn14ya0r/1/)?

**Pull request merged**